src/KMonad/Model/Button.hs

```diff
@@ -336,21 +336,25 @@ multiTap l bs = onPress $ go bs
     go []            = press l
     go ((ms, b):bs') = do
       -- This is a bit complicated. What we do is:
-      -- 1.  We wait for the release of the key that triggered this action
+      -- 1.  We wait for an event
       -- 2A. If it doesn't occur in the interval we press the button from the list
       --     and we are done.
-      -- 2B. If we do detect the release, we must now keep waiting to detect another press.
-      -- 3A. If we do not detect a press before the interval is up, we know a tap occured,
+      -- 2B. If we do detect the release of the key that triggered this action, we must now keep waiting to detect another press.
+      -- 2C. If we detect another (unrelated) press event we cancel the remaining of the multi-tap sequence and trigger a tap on the current button of the sequence.
+      -- 3A. After 2B, if we do not detect a press before the interval is up, we know a tap occured,
       --     so we tap the current button and we are done.
-      -- 3B. If we detect another press, then the user is descending into the buttons tied
+      -- 3B. If we detect another press of the same key, then the user is descending into the buttons tied
       --     to this multi-tap, so we recurse on the remaining buttons.
-      let onMatch t = do
-            within (ms - t^.elapsed) (matchMy Press)
-                   (tap b)
-                   (const $ go bs' *> pure Catch)
-            pure Catch
-      within ms (matchMy Release) (press b) onMatch
-
+      -- 3C. If we detect any other (unrelated) press event, then the multi-tap sequence is cancelled like in 2C. We trigger a tap of the current button of the sequence.
+
+      let doNext pred onTimeout next ms = tHookF InputHook ms onTimeout (\t -> do
+            pr <- pred
+            if 
+              | pr $ t^.event -> Catch <$ next (ms - t^.elapsed)
```

Comment:
Mh is this correct?  The old version called `within` twice in order to adjust the delay when we detected our own release—now we just recurse while ignoring that.

Reply:
Isn't it what happens here?

`tHookF` is called twice (through the 2 calls to `doNext`), first with `ms` as an argument (the parameter of `go`), then with `ms - t^.elapsed` where `t` is the trigger of the first hook. So the delay is adjusted in the same way as it used to be, or do I get it wrong?

Reply:
I may be being stupid, but where is `doNext` called twice?  It's certainly called once with the `ms` argument, but `next` is then defined as `\_ -> go bs')`, meaning the new delay will simply be ignored.

Reply:
`doNext` is passed as a callback to itself: in the first call, line 357, `(doNext (matchMy Press) (tap b) (\_ -> go bs'))` is passed as parameter `next` (4 arguments out of 5: time is missing).

Hence at line 353, during the first call, `next (ms - t^.elapsed)` develops to `doNext (matchMy Press) (tap b) (\_ -> go bs') (ms - t^.elapsed)`, making the second call (with corrected delay).

I hope this is correct.

Ok, even assuming this is correct, code that makes readers feel stupid can't be very good code!
`doNext` could be made into 2 functions again like in the previous version if it makes it more readable...

Reply:
Oooh of course.  `doNext` has quite a lot of arguments and I must have mixed up some parentheses there.  Sorry about that

Reply:
No worry!

Please let me know if I should do something to improve readability or if there are some coding guidelines I should be aware of.

Reply:
I think this is really fine; perhaps a type signature on `doNext` could help with understanding.  Other than that I'll try to test this myself tomorrow and then we can probably merge